File: datasets/C++/Claude/Algorithm/hard2.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

int chiefHopper(vector<int>& arr) {
    // Work backwards from the last building
    // We need at least 1 energy after completing all buildings
    int minEnergyNeeded = 1;
    
    // Go through buildings in reverse order
    for (int i = arr.size() - 1; i >= 0; i--) {
        int height = arr[i];
        // To have minEnergyNeeded after jumping, we need:
        // 2 * before_energy - height >= minEnergyNeeded
        // before_energy >= (minEnergyNeeded + height) / 2
        minEnergyNeeded = (minEnergyNeeded + height + 1) / 2; // Ceiling division
    }
    
    return minEnergyNeeded;
}

void testExample() {
    vector<int> arr = {2, 3, 4, 3, 2};
    int result = chiefHopper(arr);
    
    cout << "Example: arr = [";
    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i];
        if (i < arr.size() - 1) cout << ", ";
    }
    cout << "]" << endl;
    cout << "Minimum starting energy: " << result << endl;
    
    // Verify the solution
    int energy = result;
    cout << "\nVerification:" << endl;
    cout << "Starting energy: " << energy << endl;
    
    for (int i = 0; i < arr.size(); i++) {
        int height = arr[i];
        int newEnergy = 2 * energy - height;
        cout << "Building " << i << ": height=" << height 
             << ", energy: " << energy << " -> " << newEnergy << endl;
        energy = newEnergy;
        if (energy < 0) {
            cout << "Energy went below 0! Solution is incorrect." << endl;
            break;
        }
    }
    cout << "Final energy: " << energy << endl;
}

int main() {
    // Run test first
    testExample();
    cout << "\n" << string(50, '=') << "\n" << endl;
    
    // Main program
    int n;
    cout << "Enter the number of buildings: ";
    cin >> n;
    
    vector<int> heights(n);
    cout << "Enter the heights of buildings (space-separated): ";
    for (int i = 0; i < n; i++) {
        cin >> heights[i];
    }
    
    int result = chiefHopper(heights);
    cout << "Minimum starting energy required: " << result << endl;
    
    return 0;
}

File: datasets/Python/Claude/Algorithm/hard2.py
import math

def chiefHopper(arr):
    # Work backwards from the last building
    # We need at least 1 energy after completing all buildings
    min_energy_needed = 1
    
    # Go through buildings in reverse order
    for i in range(len(arr) - 1, -1, -1):
        height = arr[i]
        # To have min_energy_needed after jumping, we need:
        # 2 * before_energy - height >= min_energy_needed
        # before_energy >= (min_energy_needed + height) / 2
        min_energy_needed = math.ceil((min_energy_needed + height) / 2)
    
    return min_energy_needed

# Test with the given example
def test_example():
    arr = [2, 3, 4, 3, 2]
    result = chiefHopper(arr)
    print(f"Example: arr = {arr}")
    print(f"Minimum starting energy: {result}")
    
    # Verify the solution
    energy = result
    print(f"\nVerification:")
    print(f"Starting energy: {energy}")
    
    for i, height in enumerate(arr):
        new_energy = 2 * energy - height
        print(f"Building {i}: height={height}, energy: {energy} -> {new_energy}")
        energy = new_energy
        if energy < 0:
            print("Energy went below 0! Solution is incorrect.")
            break
    print(f"Final energy: {energy}")

# Main function to take input and solve
def main():
    print("Enter the number of buildings:")
    n = int(input())
    
    print("Enter the heights of buildings (space-separated):")
    heights = list(map(int, input().split()))
    
    if len(heights) != n:
        print(f"Error: Expected {n} buildings, got {len(heights)}")
        return
    
    result = chiefHopper(heights)
    print(f"Minimum starting energy required: {result}")

if __name__ == "__main__":
    # Run test first
    test_example()
    print("\n" + "="*50 + "\n")
    
    # Then run main program
    main()